File: backend/tauri/src-tauri/src/profiles.rs
//! Profile Management Module
//!
//! Manages user profiles on the desktop client. Each profile has:
//! - Isolated data storage (separate datastore.sqlite)
//! - Optional per-profile sync configuration
//! - Separate Chromium session data (on desktop)
//!
//! Profile metadata is stored in profiles.db in the userData directory.
//! Ports backend/electron/profiles.ts to Rust.

use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};

// ==================== Types ====================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Profile {
    pub id: String,
    pub name: String,
    pub slug: String,

    pub sync_enabled: bool,
    pub api_key: Option<String>,
    pub server_profile_slug: Option<String>,
    pub last_sync_at: Option<i64>,

    pub created_at: i64,
    pub last_used_at: i64,
    pub is_default: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProfileSyncConfig {
    pub api_key: String,
    pub server_profile_slug: String,
}

// ==================== Database Initialization ====================

/// Initialize the profiles database
pub fn init_profiles_db(db_path: &Path) -> rusqlite::Result<Connection> {
    let conn = Connection::open(db_path)?;
    conn.pragma_update(None, "journal_mode", "WAL")?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS profiles (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL UNIQUE,
            slug TEXT NOT NULL UNIQUE,

            sync_enabled INTEGER DEFAULT 0,
            api_key TEXT,
            server_profile_slug TEXT,
            last_sync_at INTEGER,

            created_at INTEGER NOT NULL,
            last_used_at INTEGER NOT NULL,
            is_default INTEGER DEFAULT 0
        );

        CREATE TABLE IF NOT EXISTS active_profile (
            id INTEGER PRIMARY KEY CHECK (id = 1),
            profile_slug TEXT NOT NULL
        );",
    )?;

    Ok(conn)
}

// ==================== Helpers ====================

fn row_to_profile(row: &rusqlite::Row) -> rusqlite::Result<Profile> {
    Ok(Profile {
        id: row.get("id")?,
        name: row.get("name")?,
        slug: row.get("slug")?,
        sync_enabled: row.get::<_, i64>("sync_enabled")? == 1,
        api_key: row.get("api_key")?,
        server_profile_slug: row.get("server_profile_slug")?,
        last_sync_at: row.get("last_sync_at")?,
        created_at: row.get("created_at")?,
        last_used_at: row.get("last_used_at")?,
        is_default: row.get::<_, i64>("is_default")? == 1,
    })
}

fn now() -> i64 {
    chrono::Utc::now().timestamp_millis()
}

fn generate_uuid() -> String {
    uuid::Uuid::new_v4().to_string()
}

/// Generate a filesystem-safe slug from a name
fn slugify(name: &str) -> String {
    name.to_lowercase()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join("-")
        .chars()
        .filter(|c| c.is_alphanumeric() || *c == '-')
        .collect()
}

// ==================== CRUD Operations ====================

/// List all profiles
pub fn list_profiles(conn: &Connection) -> Vec<Profile> {
    let mut stmt = conn
        .prepare("SELECT * FROM profiles ORDER BY last_used_at DESC")
        .unwrap();
    stmt.query_map([], |row| row_to_profile(row))
        .unwrap()
        .filter_map(|r| r.ok())
        .collect()
}

/// Create a new profile
pub fn create_profile(
    conn: &Connection,
    name: &str,
    user_data_path: Option<&Path>,
) -> rusqlite::Result<Profile> {
    let slug = slugify(name);

    // Check if slug already exists
    let exists: bool = conn
        .query_row(
            "SELECT COUNT(*) FROM profiles WHERE slug = ?1",
            params![slug],
            |row| row.get::<_, i64>(0),
        )
        .map(|c| c > 0)?;

    if exists {
        return Err(rusqlite::Error::QueryReturnedNoRows); // Profile exists
    }

    let id = generate_uuid();
    let timestamp = now();

    conn.execute(
        "INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default) VALUES (?1, ?2, ?3, 0, ?4, ?5, 0)",
        params![id, name, slug, timestamp, timestamp],
    )?;

    // Create profile directory
    if let Some(data_path) = user_data_path {
        let profile_dir = data_path.join(&slug);
        let _ = std::fs::create_dir_all(profile_dir);
    }

    get_profile(conn, &slug).ok_or(rusqlite::Error::QueryReturnedNoRows)
}

/// Get a profile by slug
pub fn get_profile(conn: &Connection, slug: &str) -> Option<Profile> {
    conn.query_row("SELECT * FROM profiles WHERE slug = ?1", params![slug], |row| {
        row_to_profile(row)
    })
    .ok()
}

/// Get a profile by ID
pub fn get_profile_by_id(conn: &Connection, id: &str) -> Option<Profile> {
    conn.query_row("SELECT * FROM profiles WHERE id = ?1", params![id], |row| {
        row_to_profile(row)
    })
    .ok()
}

/// Delete a profile (cannot delete default or active profile)
pub fn delete_profile(conn: &Connection, id: &str) -> Result<(), String> {
    let profile = get_profile_by_id(conn, id).ok_or("Profile not found")?;

    if profile.is_default {
        return Err("Cannot delete default profile".to_string());
    }

    let active = get_active_profile(conn);
    if active.id == id {
        return Err("Cannot delete active profile".to_string());
    }

    conn.execute("DELETE FROM profiles WHERE id = ?1", params![id])
        .map_err(|e| format!("Failed to delete profile: {}", e))?;

    Ok(())
}

// ==================== Active Profile ====================

/// Get the active profile
pub fn get_active_profile(conn: &Connection) -> Profile {
    // Try to get from active_profile table
    if let Ok(slug) = conn.query_row(
        "SELECT profile_slug FROM active_profile WHERE id = 1",
        [],
        |row| row.get::<_, String>(0),
    ) {
        if let Some(profile) = get_profile(conn, &slug) {
            return profile;
        }
    }

    // Fallback to default profile
    if let Ok(profile) = conn.query_row(
        "SELECT * FROM profiles WHERE is_default = 1",
        [],
        |row| row_to_profile(row),
    ) {
        return profile;
    }

    // Last resort: any profile
    if let Ok(profile) = conn.query_row("SELECT * FROM profiles LIMIT 1", [], |row| {
        row_to_profile(row)
    }) {
        return profile;
    }

    // Should not reach here if ensure_default_profile was called
    panic!("No profiles found. Call ensure_default_profile first.");
}

/// Set the active profile
pub fn set_active_profile(conn: &Connection, slug: &str) -> Result<(), String> {
    let profile = get_profile(conn, slug).ok_or(format!("Profile '{}' not found", slug))?;

    conn.execute(
        "INSERT OR REPLACE INTO active_profile (id, profile_slug) VALUES (1, ?1)",
        params![slug],
    )
    .map_err(|e| format!("Failed to set active profile: {}", e))?;

    // Update last_used_at
    conn.execute(
        "UPDATE profiles SET last_used_at = ?1 WHERE id = ?2",
        params![now(), profile.id],
    )
    .map_err(|e| format!("Failed to update last_used_at: {}", e))?;

    Ok(())
}

/// Ensure the default profile exists
pub fn ensure_default_profile(conn: &Connection, user_data_path: Option<&Path>) {
    let exists: bool = conn
        .query_row(
            "SELECT COUNT(*) FROM profiles WHERE slug = 'default'",
            [],
            |row| row.get::<_, i64>(0),
        )
        .map(|c| c > 0)
        .unwrap_or(false);

    if exists {
        return;
    }

    let id = generate_uuid();
    let timestamp = now();

    let _ = conn.execute(
        "INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default) VALUES (?1, 'Default', 'default', 0, ?2, ?3, 1)",
        params![id, timestamp, timestamp],
    );

    // Create profile directory
    if let Some(data_path) = user_data_path {
        let profile_dir = data_path.join("default");
        let _ = std::fs::create_dir_all(profile_dir);
    }
}

/// Migrate existing profile directories to profiles.db
pub fn migrate_existing_profiles(conn: &Connection, user_data_path: &Path) {
    let timestamp = now();

    // Check for 'default' directory
    let default_dir = user_data_path.join("default");
    if default_dir.exists() {
        let exists: bool = conn
            .query_row(
                "SELECT COUNT(*) FROM profiles WHERE slug = 'default'",
                [],
                |row| row.get::<_, i64>(0),
            )
            .map(|c| c > 0)
            .unwrap_or(false);

        if !exists {
            let id = generate_uuid();
            let _ = conn.execute(
                "INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default) VALUES (?1, 'Default', 'default', 0, ?2, ?3, 1)",
                params![id, timestamp, timestamp],
            );
            println!("[profiles] Migrated existing default profile directory");
        }
    }

    // Check for 'dev' directory
    let dev_dir = user_data_path.join("dev");
    if dev_dir.exists() {
        let exists: bool = conn
            .query_row(
                "SELECT COUNT(*) FROM profiles WHERE slug = 'dev'",
                [],
                |row| row.get::<_, i64>(0),
            )
            .map(|c| c > 0)
            .unwrap_or(false);

        if !exists {
            let id = generate_uuid();
            let _ = conn.execute(
                "INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default) VALUES (?1, 'Development', 'dev', 0, ?2, ?3, 0)",
                params![id, timestamp, timestamp],
            );
            println!("[profiles] Migrated existing dev profile directory");
        }
    }

    // Ensure at least default profile exists
    ensure_default_profile(conn, Some(user_data_path));
}

// ==================== Sync Configuration ====================

/// Enable sync for a profile
pub fn enable_sync(
    conn: &Connection,
    profile_id: &str,
    api_key: &str,
    server_profile_slug: &str,
) -> Result<(), String> {
    let _profile = get_profile_by_id(conn, profile_id).ok_or("Profile not found")?;

    conn.execute(
        "UPDATE profiles SET sync_enabled = 1, api_key = ?1, server_profile_slug = ?2 WHERE id = ?3",
        params![api_key, server_profile_slug, profile_id],
    )
    .map_err(|e| format!("Failed to enable sync: {}", e))?;

    Ok(())
}

/// Disable sync for a profile
pub fn disable_sync(conn: &Connection, profile_id: &str) -> Result<(), String> {
    conn.execute(
        "UPDATE profiles SET sync_enabled = 0, api_key = NULL, server_profile_slug = NULL, last_sync_at = NULL WHERE id = ?1",
        params![profile_id],
    )
    .map_err(|e| format!("Failed to disable sync: {}", e))?;

    Ok(())
}

/// Get sync configuration for a profile
pub fn get_sync_config(conn: &Connection, profile_id: &str) -> Option<ProfileSyncConfig> {
    let profile = get_profile_by_id(conn, profile_id)?;

    if !profile.sync_enabled {
        return None;
    }

    let api_key = profile.api_key?;
    let server_profile_slug = profile.server_profile_slug?;

    Some(ProfileSyncConfig {
        api_key,
        server_profile_slug,
    })
}

/// Update last sync time for a profile
pub fn update_last_sync_time(
    conn: &Connection,
    profile_id: &str,
    timestamp: i64,
) -> Result<(), String> {
    conn.execute(
        "UPDATE profiles SET last_sync_at = ?1 WHERE id = ?2",
        params![timestamp, profile_id],
    )
    .map_err(|e| format!("Failed to update last_sync_at: {}", e))?;

    Ok(())
}

File: backend/electron/profiles.ts
/**
 * Profile Management Module
 *
 * Manages user profiles on the desktop client. Each profile has:
 * - Isolated data storage (separate datastore.sqlite)
 * - Optional per-profile sync configuration
 * - Separate Chromium session data (on desktop)
 *
 * Profile metadata is stored in profiles.db in the userData directory.
 */

import Database from 'better-sqlite3';
import path from 'node:path';
import fs from 'node:fs';
import crypto from 'node:crypto';

export interface Profile {
  id: string;                   // UUID
  name: string;                 // User-visible name (e.g., "Work")
  folder: string;               // Filesystem-safe folder name (e.g., "work")

  // Sync configuration (optional)
  syncEnabled: boolean;
  apiKey: string | null;        // Server user API key
  serverProfileId: string | null;  // Which server profile UUID to sync to
  lastSyncAt: number | null;    // Unix ms

  createdAt: number;            // Unix ms
  lastUsedAt: number;           // Unix ms
  isDefault: boolean;           // Cannot be deleted
}

export interface SyncConfig {
  apiKey: string;
  serverProfileId: string;
}

let profilesDb: Database.Database | null = null;
let userDataPath: string | null = null;

/**
 * Initialize the profiles database
 * @param dataPath - Base userData directory
 * @param filename - Database filename (default: 'profiles.db', dev: '.dev-profiles.db')
 */
export function initProfilesDb(dataPath: string, filename: string = 'profiles.db'): Database.Database {
  userDataPath = dataPath;
  const profilesDbPath = path.join(dataPath, filename);

  profilesDb = new Database(profilesDbPath);
  profilesDb.pragma('journal_mode = WAL');

  // Create profiles table
  profilesDb.exec(`
    CREATE TABLE IF NOT EXISTS profiles (
      id TEXT PRIMARY KEY,
      name TEXT NOT NULL UNIQUE,
      slug TEXT NOT NULL UNIQUE,

      sync_enabled INTEGER DEFAULT 0,
      api_key TEXT,
      server_profile_slug TEXT,
      last_sync_at INTEGER,

      created_at INTEGER NOT NULL,
      last_used_at INTEGER NOT NULL,
      is_default INTEGER DEFAULT 0
    );

    CREATE TABLE IF NOT EXISTS active_profile (
      id INTEGER PRIMARY KEY CHECK (id = 1),
      profile_slug TEXT NOT NULL
    );
  `);

  return profilesDb;
}

/**
 * Get the profiles database (must be initialized first)
 */
function getProfilesDb(): Database.Database {
  if (!profilesDb) {
    throw new Error('Profiles database not initialized. Call initProfilesDb first.');
  }
  return profilesDb;
}

/**
 * Convert database row to Profile object
 */
function rowToProfile(row: any): Profile {
  return {
    id: row.id,
    name: row.name,
    folder: row.slug,
    syncEnabled: row.sync_enabled === 1,
    apiKey: row.api_key || null,
    serverProfileId: row.server_profile_slug || null,
    lastSyncAt: row.last_sync_at || null,
    createdAt: row.created_at,
    lastUsedAt: row.last_used_at,
    isDefault: row.is_default === 1,
  };
}

/**
 * List all profiles
 */
export function listProfiles(): Profile[] {
  const db = getProfilesDb();
  const rows = db.prepare(`
    SELECT * FROM profiles
    ORDER BY last_used_at DESC
  `).all();

  return rows.map(rowToProfile);
}

/**
 * Create a new profile
 */
export function createProfile(name: string): Profile {
  const db = getProfilesDb();

  // Generate folder name from name (lowercase, replace spaces with hyphens, remove special chars)
  const folder = name.toLowerCase()
    .replace(/\s+/g, '-')
    .replace(/[^a-z0-9-]/g, '');

  // Check if folder already exists
  const existing = db.prepare('SELECT id FROM profiles WHERE slug = ?').get(folder);
  if (existing) {
    throw new Error(`Profile with folder '${folder}' already exists`);
  }

  const id = crypto.randomUUID();
  const timestamp = Date.now();

  db.prepare(`
    INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default)
    VALUES (?, ?, ?, 0, ?, ?, 0)
  `).run(id, name, folder, timestamp, timestamp);

  // Create profile directory
  if (userDataPath) {
    const profileDir = path.join(userDataPath, folder);
    if (!fs.existsSync(profileDir)) {
      fs.mkdirSync(profileDir, { recursive: true });
    }
  }

  return getProfileByFolder(folder)!;
}

/**
 * Get a specific profile by folder name
 */
export function getProfileByFolder(folder: string): Profile | null {
  const db = getProfilesDb();
  const row = db.prepare('SELECT * FROM profiles WHERE slug = ?').get(folder);
  return row ? rowToProfile(row) : null;
}

/**
 * Get a specific profile by ID
 */
export function getProfileById(id: string): Profile | null {
  const db = getProfilesDb();
  const row = db.prepare('SELECT * FROM profiles WHERE id = ?').get(id);
  return row ? rowToProfile(row) : null;
}

/**
 * Update a profile
 */
export function updateProfile(id: string, updates: Partial<Profile>): void {
  const db = getProfilesDb();

  const allowedUpdates: Array<keyof Profile> = ['name', 'lastUsedAt'];
  const setClauses: string[] = [];
  const values: any[] = [];

  for (const key of allowedUpdates) {
    if (updates[key] !== undefined) {
      const dbKey = key === 'lastUsedAt' ? 'last_used_at' : key;
      setClauses.push(`${dbKey} = ?`);
      values.push(updates[key]);
    }
  }

  if (setClauses.length === 0) {
    return;
  }

  values.push(id);

  db.prepare(`
    UPDATE profiles SET ${setClauses.join(', ')} WHERE id = ?
  `).run(...values);
}

/**
 * Delete a profile (cannot delete default or active profile)
 */
export function deleteProfile(id: string): void {
  const db = getProfilesDb();

  const profile = getProfileById(id);
  if (!profile) {
    throw new Error('Profile not found');
  }

  if (profile.isDefault) {
    throw new Error('Cannot delete default profile');
  }

  const activeProfile = getActiveProfile();
  if (activeProfile.id === id) {
    throw new Error('Cannot delete active profile');
  }

  db.prepare('DELETE FROM profiles WHERE id = ?').run(id);

  // Note: Profile directory is NOT deleted - data is preserved
  // User can manually delete the directory if desired
}

/**
 * Get the active profile
 */
export function getActiveProfile(): Profile {
  const db = getProfilesDb();

  // Try to get from active_profile table
  const activeRow = db.prepare('SELECT profile_slug FROM active_profile WHERE id = 1').get() as { profile_slug: string } | undefined;

  if (activeRow) {
    const profile = getProfileByFolder(activeRow.profile_slug);
    if (profile) {
      return profile;
    }
  }

  // Fallback to default profile
  const defaultProfile = db.prepare('SELECT * FROM profiles WHERE is_default = 1').get();
  if (defaultProfile) {
    return rowToProfile(defaultProfile);
  }

  // Last resort: return any profile
  const anyProfile = db.prepare('SELECT * FROM profiles LIMIT 1').get();
  if (anyProfile) {
    return rowToProfile(anyProfile);
  }

  throw new Error('No profiles found. Call ensureDefaultProfile first.');
}

/**
 * Set the active profile (persists across app restarts)
 */
export function setActiveProfile(folder: string): void {
  const db = getProfilesDb();

  const profile = getProfileByFolder(folder);
  if (!profile) {
    throw new Error(`Profile '${folder}' not found`);
  }

  // Update active profile
  db.prepare(`
    INSERT OR REPLACE INTO active_profile (id, profile_slug)
    VALUES (1, ?)
  `).run(folder);

  // Update last_used_at
  updateProfile(profile.id, { lastUsedAt: Date.now() });
}

/**
 * Ensure the default profile exists
 */
export function ensureDefaultProfile(): void {
  const db = getProfilesDb();

  const existing = db.prepare('SELECT id FROM profiles WHERE slug = ?').get('default');
  if (existing) {
    return;
  }

  const id = crypto.randomUUID();
  const timestamp = Date.now();

  db.prepare(`
    INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default)
    VALUES (?, 'Default', 'default', 0, ?, ?, 1)
  `).run(id, timestamp, timestamp);

  // Create profile directory
  if (userDataPath) {
    const profileDir = path.join(userDataPath, 'default');
    if (!fs.existsSync(profileDir)) {
      fs.mkdirSync(profileDir, { recursive: true });
    }
  }
}

/**
 * Migrate existing profile directories to profiles.db
 * Detects 'default' and 'dev' directories and creates profile records if they don't exist
 */
export function migrateExistingProfiles(): void {
  if (!userDataPath) {
    throw new Error('userDataPath not set. Call initProfilesDb first.');
  }

  const db = getProfilesDb();
  const timestamp = Date.now();

  // Check for 'default' directory
  const defaultDir = path.join(userDataPath, 'default');
  if (fs.existsSync(defaultDir)) {
    const existing = db.prepare('SELECT id FROM profiles WHERE slug = ?').get('default');
    if (!existing) {
      const id = crypto.randomUUID();
      db.prepare(`
        INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default)
        VALUES (?, 'Default', 'default', 0, ?, ?, 1)
      `).run(id, timestamp, timestamp);
      console.log('[profiles] Migrated existing default profile directory');
    }
  }

  // Check for 'dev' directory
  const devDir = path.join(userDataPath, 'dev');
  if (fs.existsSync(devDir)) {
    const existing = db.prepare('SELECT id FROM profiles WHERE slug = ?').get('dev');
    if (!existing) {
      const id = crypto.randomUUID();
      db.prepare(`
        INSERT INTO profiles (id, name, slug, sync_enabled, created_at, last_used_at, is_default)
        VALUES (?, 'Development', 'dev', 0, ?, ?, 0)
      `).run(id, timestamp, timestamp);
      console.log('[profiles] Migrated existing dev profile directory');
    }
  }

  // Ensure at least default profile exists
  ensureDefaultProfile();
}

// ==================== Sync Configuration ====================

/**
 * Enable sync for a profile
 */
export function enableSync(profileId: string, apiKey: string, serverProfileId: string): void {
  const db = getProfilesDb();

  const profile = getProfileById(profileId);
  if (!profile) {
    throw new Error('Profile not found');
  }

  db.prepare(`
    UPDATE profiles
    SET sync_enabled = 1, api_key = ?, server_profile_slug = ?
    WHERE id = ?
  `).run(apiKey, serverProfileId, profileId);
}

/**
 * Disable sync for a profile
 */
export function disableSync(profileId: string): void {
  const db = getProfilesDb();

  db.prepare(`
    UPDATE profiles
    SET sync_enabled = 0, api_key = NULL, server_profile_slug = NULL, last_sync_at = NULL
    WHERE id = ?
  `).run(profileId);
}

/**
 * Get sync configuration for a profile
 */
export function getSyncConfig(profileId: string): SyncConfig | null {
  const profile = getProfileById(profileId);
  if (!profile || !profile.syncEnabled || !profile.apiKey || !profile.serverProfileId) {
    return null;
  }

  return {
    apiKey: profile.apiKey,
    serverProfileId: profile.serverProfileId,
  };
}

/**
 * Update last sync time for a profile
 */
export function updateLastSyncTime(profileId: string, timestamp: number): void {
  const db = getProfilesDb();

  db.prepare(`
    UPDATE profiles SET last_sync_at = ? WHERE id = ?
  `).run(timestamp, profileId);
}

/**
 * Close the profiles database
 */
export function closeProfilesDb(): void {
  if (profilesDb) {
    profilesDb.close();
    profilesDb = null;
  }
}
